File: src/test/java/org/openrewrite/java/migrate/javax/AddJaxwsDependenciesTest.java
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.migrate.javax;

import org.junit.jupiter.api.Test;
import org.openrewrite.config.Environment;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.assertj.core.api.Assertions.assertThat;
import static org.openrewrite.gradle.Assertions.buildGradle;
import static org.openrewrite.gradle.toolingapi.Assertions.withToolingApi;
import static org.openrewrite.maven.Assertions.pomXml;

class AddJaxwsDependenciesTest implements RewriteTest {

    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(Environment.builder()
          .scanRuntimeClasspath("org.openrewrite.java.migrate.javax")
          .build()
          .activateRecipes("org.openrewrite.java.migrate.javax.AddJaxwsDependencies"));
    }

    @Test
    void addJaxwsRuntimeOnce() {
        rewriteRun(
          spec -> spec.beforeRecipe(withToolingApi()),
          buildGradle(
            //language=gradle
            """
              plugins {
                  id "java-library"
              }
               
              repositories {
                  mavenCentral()
              }
               
              dependencies {
                  implementation "jakarta.xml.ws:jakarta.xml.ws-api:2.3.2"
              }
              """,
            spec -> spec.after(buildGradle -> {
                Matcher version = Pattern.compile("2\\.\\d+(\\.\\d+)?").matcher(buildGradle);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                //language=gradle
                return """
                  plugins {
                      id "java-library"
                  }
                   
                  repositories {
                      mavenCentral()
                  }
                   
                  dependencies {
                      compileOnly "com.sun.xml.ws:jaxws-rt:%s"
                                    
                      implementation "jakarta.xml.ws:jakarta.xml.ws-api:%s"
                                    
                      testImplementation "com.sun.xml.ws:jaxws-rt:%s"
                  }
                  """.formatted(rtVersion, wsApiVersion, rtVersion);
            })
          ),
          pomXml(
            //language=xml
            """
              <project>
                  <groupId>com.example.jaxws</groupId>
                  <artifactId>jaxws-example</artifactId>
                  <version>1.0.0</version>
                  <dependencies>
                      <dependency>
                          <groupId>jakarta.xml.ws</groupId>
                          <artifactId>jakarta.xml.ws-api</artifactId>
                          <version>2.3.2</version>
                      </dependency>
                  </dependencies>
              </project>
              """,
            spec -> spec.after(pom -> {
                Matcher version = Pattern.compile("2.\\d+(.\\d+)?").matcher(pom);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                //language=xml
                return """
                  <project>
                      <groupId>com.example.jaxws</groupId>
                      <artifactId>jaxws-example</artifactId>
                      <version>1.0.0</version>
                      <dependencies>
                          <dependency>
                              <groupId>jakarta.xml.ws</groupId>
                              <artifactId>jakarta.xml.ws-api</artifactId>
                              <version>%s</version>
                          </dependency>
                          <dependency>
                              <groupId>com.sun.xml.ws</groupId>
                              <artifactId>jaxws-rt</artifactId>
                              <version>%s</version>
                              <scope>provided</scope>
                          </dependency>
                      </dependencies>
                  </project>
                  """.formatted(wsApiVersion, rtVersion);
            })
          )
        );
    }

    @Test
    void removeReferenceImplementationRuntime() {
        rewriteRun(
          spec -> spec.beforeRecipe(withToolingApi()),
          buildGradle(
            //language=gradle
            """
              plugins {
                  id "java-library"
              }
                            
              repositories {
                  mavenCentral()
              }
                            
              dependencies {
                  compileOnly "com.sun.xml.ws:jaxws-ri:2.3.2"
                            
                  implementation "javax.xml.ws:jaxws-api:2.3.1"
                            
                  testImplementation "com.sun.xml.ws:jaxws-ri:2.3.2"
              }
              """,
            spec -> spec.after(buildGradle -> {
                Matcher version = Pattern.compile("2\\.\\d+(\\.\\d+)?").matcher(buildGradle);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                //language=gradle
                return """
                  plugins {
                      id "java-library"
                  }
                                    
                  repositories {
                      mavenCentral()
                  }
                                    
                  dependencies {
                      implementation "com.sun.xml.ws:jaxws-rt:%s"
                                    
                      implementation "jakarta.xml.ws:jakarta.xml.ws-api:%s"
                  }
                  """.formatted(rtVersion, wsApiVersion);
            })
          ),
          pomXml(
            //language=xml
            """
              <project>
                  <groupId>com.example.jaxws</groupId>
                  <artifactId>jaxws-example</artifactId>
                  <version>1.0.0</version>
                  <dependencies>
                      <dependency>
                          <groupId>javax.xml.ws</groupId>
                          <artifactId>jaxws-api</artifactId>
                          <version>2.3.1</version>
                      </dependency>
                      <dependency>
                          <groupId>com.sun.xml.ws</groupId>
                          <artifactId>jaxws-ri</artifactId>
                          <version>2.3.2</version>
                          <scope>provided</scope>
                      </dependency>
                  </dependencies>
              </project>
              """,
            spec -> spec.after(pom -> {
                Matcher version = Pattern.compile("2.\\d+(.\\d+)?").matcher(pom);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                //language=xml
                return """
                  <project>
                      <groupId>com.example.jaxws</groupId>
                      <artifactId>jaxws-example</artifactId>
                      <version>1.0.0</version>
                      <dependencies>
                          <dependency>
                              <groupId>jakarta.xml.ws</groupId>
                              <artifactId>jakarta.xml.ws-api</artifactId>
                              <version>%s</version>
                          </dependency>
                          <dependency>
                              <groupId>com.sun.xml.ws</groupId>
                              <artifactId>jaxws-rt</artifactId>
                              <version>%s</version>
                              <scope>provided</scope>
                          </dependency>
                      </dependencies>
                  </project>
                  """.formatted(wsApiVersion, rtVersion);
            })
          )
        );
    }

    @Test
    void renameAndUpdateApiAndRuntime() {
        rewriteRun(
          spec -> spec.beforeRecipe(withToolingApi()),
          buildGradle(
            //language=gradle
            """
              plugins {
                  id "java-library"
              }
                            
              repositories {
                  mavenCentral()
              }
                            
              dependencies {
                  compileOnly "com.sun.xml.ws:jaxws-ri:2.3.2"
                            
                  implementation "jakarta.xml.ws:jakarta.xml.ws-api:2.3.2"
                            
                  testImplementation "com.sun.xml.ws:jaxws-ri:2.3.2"
              }
              """,
            spec -> spec.after(buildGradle -> {
                Matcher version = Pattern.compile("2\\.\\d+(\\.\\d+)?").matcher(buildGradle);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                //language=gradle
                return """
                  plugins {
                      id "java-library"
                  }
                                    
                  repositories {
                      mavenCentral()
                  }
                                    
                  dependencies {
                      implementation "com.sun.xml.ws:jaxws-rt:%s"
                                    
                      implementation "jakarta.xml.ws:jakarta.xml.ws-api:%s"
                  }
                  """.formatted(rtVersion, wsApiVersion);
            })
          ),
          pomXml(
            //language=xml
            """
              <project>
                  <groupId>com.example.jaxws</groupId>
                  <artifactId>jaxws-example</artifactId>
                  <version>1.0.0</version>
                  <dependencies>
                      <dependency>
                          <groupId>jakarta.xml.ws</groupId>
                          <artifactId>jakarta.xml.ws-api</artifactId>
                          <version>2.3.2</version>
                      </dependency>
                      <dependency>
                          <groupId>com.sun.xml.ws</groupId>
                          <artifactId>jaxws-ri</artifactId>
                          <version>2.3.2</version>
                          <scope>provided</scope>
                      </dependency>
                  </dependencies>
              </project>
              """,
            spec -> spec.after(pom -> {
                Matcher version = Pattern.compile("2.\\d+(.\\d+)?").matcher(pom);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                //language=xml
                return """
                  <project>
                      <groupId>com.example.jaxws</groupId>
                      <artifactId>jaxws-example</artifactId>
                      <version>1.0.0</version>
                      <dependencies>
                          <dependency>
                              <groupId>jakarta.xml.ws</groupId>
                              <artifactId>jakarta.xml.ws-api</artifactId>
                              <version>%s</version>
                          </dependency>
                          <dependency>
                              <groupId>com.sun.xml.ws</groupId>
                              <artifactId>jaxws-rt</artifactId>
                              <version>%s</version>
                              <scope>provided</scope>
                          </dependency>
                      </dependencies>
                  </project>
                  """.formatted(wsApiVersion, rtVersion);
            })
          )
        );
    }

    @Test
    void renameAndUpdateApiAndAddRuntimeManagedDependencies() {
        rewriteRun(
          spec -> spec.beforeRecipe(withToolingApi()),
          buildGradle(
            //language=gradle
            """
              plugins {
                  id "java-library"
              }
                            
              repositories {
                  mavenCentral()
              }
                            
              dependencies {
                  implementation "jakarta.xml.ws:jakarta.xml.ws-api:2.3.2"
              }
              """,
            spec -> spec.after(buildGradle -> {
                Matcher version = Pattern.compile("2\\.\\d+(\\.\\d+)?").matcher(buildGradle);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                //language=gradle
                return """
                  plugins {
                      id "java-library"
                  }
                                    
                  repositories {
                      mavenCentral()
                  }
                                    
                  dependencies {
                      compileOnly "com.sun.xml.ws:jaxws-rt:%s"
                                    
                      implementation "jakarta.xml.ws:jakarta.xml.ws-api:%s"
                                    
                      testImplementation "com.sun.xml.ws:jaxws-rt:%s"
                  }
                  """.formatted(rtVersion, wsApiVersion, rtVersion);
            })
          ),
          pomXml(
            //language=xml
            """
              <project>
                  <groupId>com.example.jaxws</groupId>
                  <artifactId>jaxws-example</artifactId>
                  <version>1.0.0</version>
                  <dependencyManagement>
                      <dependencies>
                          <dependency>
                              <groupId>javax.xml.ws</groupId>
                              <artifactId>jaxws-api</artifactId>
                              <version>2.3.1</version>
                          </dependency>
                          <dependency>
                              <groupId>com.sun.xml.ws</groupId>
                              <artifactId>jaxws-ri</artifactId>
                              <version>2.3.2</version>
                          </dependency>
                      </dependencies>
                  </dependencyManagement>
                  <dependencies>
                      <dependency>
                          <groupId>javax.xml.ws</groupId>
                          <artifactId>jaxws-api</artifactId>
                      </dependency>
                  </dependencies>
              </project>
              """,
            spec -> spec.after(pom -> {
                Matcher version = Pattern.compile("2.\\d+(.\\d+)?").matcher(pom);
                assertThat(version.find()).isTrue();
                String wsApiVersion = version.group(0);
                assertThat(version.find()).isTrue();
                String rtVersion = version.group(0);
                //language=xml
                return """
                  <project>
                      <groupId>com.example.jaxws</groupId>
                      <artifactId>jaxws-example</artifactId>
                      <version>1.0.0</version>
                      <dependencyManagement>
                          <dependencies>
                              <dependency>
                                  <groupId>jakarta.xml.ws</groupId>
                                  <artifactId>jakarta.xml.ws-api</artifactId>
                                  <version>%s</version>
                              </dependency>
                          </dependencies>
                      </dependencyManagement>
                      <dependencies>
                          <dependency>
                              <groupId>jakarta.xml.ws</groupId>
                              <artifactId>jakarta.xml.ws-api</artifactId>
                          </dependency>
                          <dependency>
                              <groupId>com.sun.xml.ws</groupId>
                              <artifactId>jaxws-rt</artifactId>
                              <version>%s</version>
                              <scope>provided</scope>
                          </dependency>
                      </dependencies>
                  </project>
                  """.formatted(wsApiVersion, rtVersion);
            })
          )
        );
    }
}

File: src/test/kotlin/org/openrewrite/java/migrate/javax/AddJaxwsDependenciesTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.migrate.javax

import org.junit.jupiter.api.Test
import org.openrewrite.Recipe
import org.openrewrite.config.Environment
import org.openrewrite.maven.MavenRecipeTest

class AddJaxwsDependenciesTest : MavenRecipeTest {
    override val recipe: Recipe = Environment.builder()
        .scanRuntimeClasspath("org.openrewrite.java.migrate")
        .build()
        .activateRecipes("org.openrewrite.java.migrate.javax.AddJaxwsDependencies")

    companion object {
        private val jaxbContextStub : String = """
            package javax.jws;

            import java.lang.annotation.Target;
            import java.lang.annotation.Retention;
            import java.lang.annotation.RetentionPolicy;
            import java.lang.annotation.ElementType;

            @Retention(value = RetentionPolicy.RUNTIME)
            @Target(value = {ElementType.TYPE})
            public @interface WebService {
            }
        """.trimIndent()

        private val javaSourceWithJaxws : String = """
            package org.old.code;
            import javax.jws.WebService;
            
            @WebService
            public class Yep {
            }
        """.trimIndent()
        private val javaSourceWithoutJaxws : String = """
            package org.old.code;
            public class Nope {
            }
        """.trimIndent()
    }

    @Test
    fun onlyIfUsingJaxws() = assertChanged(
        before = """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
              </dependencies>
            </project>
        """.trimIndent(),
        after ="""
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
                <dependency>
                  <groupId>jakarta.xml.ws</groupId>
                  <artifactId>jakarta.xml.ws-api</artifactId>
                  <version>2.3.3</version>
                </dependency>
                <dependency>
                  <groupId>com.sun.xml.ws</groupId>
                  <artifactId>jaxws-rt</artifactId>
                  <version>2.3.4</version>
                  <scope>runtime</scope>
                </dependency>
              </dependencies>
            </project>
        """.trimIndent(),
        additionalJavaFiles = arrayOf(jaxbContextStub, javaSourceWithJaxws)
    )

    @Test
    fun doNotChangeIfNoJaxws() = assertUnchanged(
        before = """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
              </dependencies>
            </project>
        """.trimIndent(),
        additionalJavaFiles = arrayOf(javaSourceWithoutJaxws)
    )

    @Test
    fun migrateFromJaxwsApiToJakarta() = assertChanged(
        before = """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
                <dependency>
                  <groupId>javax.xml.ws</groupId>
                  <artifactId>jaxws-api</artifactId>
                  <version>2.3.1</version>
                </dependency>
                <dependency>
                  <groupId>com.sun.xml.ws</groupId>
                  <artifactId>jaxws-rt</artifactId>
                  <version>2.3.2</version>
                  <scope>runtime</scope>
                </dependency>
              </dependencies>
            </project>
        """.trimIndent(),
        after ="""
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
                <dependency>
                  <groupId>jakarta.xml.ws</groupId>
                  <artifactId>jakarta.xml.ws-api</artifactId>
                  <version>2.3.3</version>
                </dependency>
                <dependency>
                  <groupId>com.sun.xml.ws</groupId>
                  <artifactId>jaxws-rt</artifactId>
                  <version>2.3.4</version>
                  <scope>runtime</scope>
                </dependency>
              </dependencies>
            </project>
        """.trimIndent(),
        additionalJavaFiles = arrayOf(jaxbContextStub, javaSourceWithJaxws)
    )
    @Test
    fun migrateFromJaxwsToJakartaParent() = assertChanged(
        before = """
            <project>
                <parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>my-parent</artifactId>
                    <version>1</version>
                </parent>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-child</artifactId>
                <version>1</version>
                <dependencies>
                    <dependency>
                        <groupId>javax.xml.ws</groupId>
                        <artifactId>jaxws-api</artifactId>
                    </dependency>
                    <dependency>
                        <groupId>com.sun.xml.ws</groupId>
                        <artifactId>jaxws-rt</artifactId>
                        <scope>runtime</scope>
                    </dependency>
                </dependencies>
            </project>
        """.trimIndent(),
        after ="""
            <project>
                <parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>my-parent</artifactId>
                    <version>1</version>
                </parent>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-child</artifactId>
                <version>1</version>
                <dependencies>
                    <dependency>
                        <groupId>jakarta.xml.ws</groupId>
                        <artifactId>jakarta.xml.ws-api</artifactId>
                        <version>2.3.3</version>
                    </dependency>
                    <dependency>
                        <groupId>com.sun.xml.ws</groupId>
                        <artifactId>jaxws-rt</artifactId>
                        <scope>runtime</scope>
                    </dependency>
                </dependencies>
            </project>
        """.trimIndent(),
        additionalJavaFiles = arrayOf(jaxbContextStub, javaSourceWithJaxws),
        additionalMavenFiles = arrayOf(
            """
            <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-parent</artifactId>
                <version>1</version>
                <dependencyManagement>
                    <dependencies>
                        <dependency>
                            <groupId>javax.xml.ws</groupId>
                            <artifactId>jaxws-api</artifactId>
                            <version>2.3.1</version>
                        </dependency>
                        <dependency>
                            <groupId>com.sun.xml.ws</groupId>
                            <artifactId>jaxws-rt</artifactId>
                            <version>2.3.2</version>
                            <scope>runtime</scope>
                        </dependency>
                      </dependencies>
                </dependencyManagement>
            </project>
            """.trimIndent()
        )
    )

}
